### PR TITLE
Test conflicting tx situations.

### LINKED ISSUE #958: More tests for `IndexedTxGraph` and `TxGraph`
**Describe the enhancement**  

Now that #926 is merged, let's look at having more fine-grained tests.

Please comment below if you wish to work on a test (to avoid duplicate work!).

### For `TxGraph`

- [ ] Test behavior of inserting `Anchor`s without the corresponding transaction existing in the graph. We should allow for this behavior.
    1. Insert various anchors for a non-existing `txid`. Ensure `Additions` are as expected.
    2. Merge the resultant `Additions` of the previous step, and attempt to recover a new `TxGraph` with them (this should succeed).
    3. Now, insert the actual transaction of `txid`. Ensure anchors still exist.
    4. Now, attempt recovery of the `TxGraph` again with the appended additions.
    - @remix75 

- [ ] Test `TxGraph::get_chain_position` and `TxGraph::list_chain_txs` behavior with unconfirmed conflicts. "Unconfirmed" can either mean that the txs do not have anchors, or that the anchors are not in the best chain. Both cases need to be tested. `TxGraph` should always pick the transaction with the latest `last_seen` value. What happens if the conflicts have the same `last_seen` value? - @rajarshimaitra

- [ ] Test `TxGraph::get_chain_position` and `TxGraph::list_chain_txs` behavior with unconfirmed transactions (txs with no anchors) that conflict with transactions that are confirmed. The unconfirmed transactions should not be givne a chain position or be listed. - @LagginTimes

### PR DESCRIPTION
**Pull request by rajarshimaitra**
### Description

partial fixes #958 . These tests attempt to cover all the conflicting scenarios. 

The following tests are included:

 - Mempool conflicts are chosen by the latest last seen.
 - Conflicts in mempool and orphaned block filters out the orphaned block tx.
 - Conflicts with mempool and confirmed tx, filters out mempool.
 - Conflicts with two mempool tx at same `seen_at`.

 Some helper functions in the common module.
### Notes to the reviewers

As noted by @evanlinjin on the test description.
> What happens if the conflicts have the same last_seen value?

A possible solution for this situation is mentioned in the comment of the last test, which covers this case. I will attempt a PR to enforce the behavior if that is a reasonable approach. 

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

### CODE REVIEW

**Comment by notmandatory** [comment]
Looks like this one needs a `cargo fmt` .. otherwise is it ready to review or dependent on #963 ?

**Comment by evanlinjin** [comment]
> Looks like this one needs a `cargo fmt` .. otherwise is it ready to review or dependent on #963 ?

No dependencies, I saw that CI wasn't passing so didn't review.

**Comment by rajarshimaitra** [comment]
My bad. Had the fmt fix, but forgot to push. This had a dependency on #963 as the Anchor type changed. Fixed now..

**Comment by evanlinjin** [comment]
Will look at this first thing tomorrow.

File: crates/chain/tests/test_indexed_tx_graph.rs [hunk 1]
```diff
@@ -452,3 +452,264 @@ fn test_list_owned_txouts() {
...
+            ObservedAs::Unconfirmed(200)
+        ]
+        .into()
+    );
+
+    // We only have one utxo. The latest `tx_conflict_2`.
+    assert_eq!(
+        utxos.pop().unwrap().chain_position,
+        ObservedAs::Unconfirmed(200)
+    );
```
Comment (evanlinjin) [inline crates/chain/tests/test_indexed_tx_graph.rs:510]: Why could we not `assert` all of `utxo`?

Comment (rajarshimaitra) [reply crates/chain/tests/test_indexed_tx_graph.rs:510]: We can, but will make the tests code too verbose. The point here is to test various txout positions. If we wanna test the whole structure of `FullTxOut`, could be done in a separate test.

**Comment by evanlinjin** [comment]
@rajarshimaitra as we discussed.

```rust
struct TxTemplate<'a, K> {
    pub tx_name: &'a str,
    pub spends: &'a [(&'a str, usize)], // (tx_name, vout)
    pub outputs: &'a [TxOutTemplate<K>],
    pub anchors: &'a [(u32, BlockId)],
    pub last_seen: Option<u64>,
}

struct TxOutTemplate<K> {
    pub value: u64,
    pub owned_spk: Option<(K, u32)>, // some = derive a spk from K, none = random spk
}

fn init_graph<'a, K: 'a>(
    descriptors: HashMap<K, Descriptor<DescriptorPublicKey>>,
    tx_templates: impl IntoIterator<Item = TxTemplate<'a, K>>,
) -> (
    IndexedTxGraph<ConfirmationHeightAnchor, KeychainTxOutIndex<K>>,
    HashMap<&'a str, Txid>,
) {
    todo!()
}

#[test]
fn example_test() {
    struct TestCase<'a, K> {
        tx_templates: &'a [TxTemplate<'a, K>],
        chain_tip: BlockId,
        expected_list_chain_txs: (),
    }
}
```

**Comment by evanlinjin** [comment]
@LagginTimes Could you take over on this?

**Comment by LagginTimes** [comment]
> @LagginTimes Could you take over on this?

Sure, I can start looking into this.

**Comment by notmandatory** [comment]
@LagginTimes any updates on this? should I keep it in the alpha.1 milestone?

**Comment by evanlinjin** [comment]
Replaced by #1064

**Pull request closed without merging**